llvm/include/llvm/ADT/PagedVector.h

```diff
@@ -0,0 +1,301 @@
+class PagedVector {
+  static_assert(PAGE_SIZE > 0, "PAGE_SIZE must be greater than 0. Most likely "
+                               "you want it to be greater than 16.");
+  // The actual number of element in the vector which can be accessed.
+  std::size_t Size = 0;
+
+  // The position of the initial element of the page in the Data vector.
+  // Pages are allocated contiguously in the Data vector.
+  mutable std::vector<T *> PageToDataIdx;
+  // Actual page data. All the page elements are added to this vector on the
+  // first access of any of the elements of the page. Elements default
+  // constructed and elements of the page are stored contiguously. The order of
+  // the elements however depends on the order of access of the pages.
+  uintptr_t Allocator = 0;
+
+  constexpr static T *invalidPage() { return reinterpret_cast<T *>(SIZE_MAX); }
+
+public:
+  // Default constructor. We build our own allocator.
+  PagedVector()
+      : Allocator(reinterpret_cast<uintptr_t>(new BumpPtrAllocator) | 0x1) {}
+  PagedVector(BumpPtrAllocator *A)
+      : Allocator(reinterpret_cast<uintptr_t>(A)) {}
+
+  ~PagedVector() {
+    // If we own the allocator, delete it.
+    if (Allocator & 0x1) {
+      delete getAllocator();
+    }
+  }
+
+  // Get the allocator.
+  BumpPtrAllocator *getAllocator() const {
+    return reinterpret_cast<BumpPtrAllocator *>(Allocator & ~0x1);
+  }
+  // Lookup an element at position Index.
+  T &operator[](std::size_t Index) const { return at(Index); }
+
+  // Lookup an element at position i.
+  // If the associated page is not filled, it will be filled with default
+  // constructed elements. If the associated page is filled, return the element.
+  T &at(std::size_t Index) const {
+    assert(Index < Size);
+    assert(Index / PAGE_SIZE < PageToDataIdx.size());
+    auto *&PagePtr = PageToDataIdx[Index / PAGE_SIZE];
+    // If the page was not yet allocated, allocate it.
+    if (PagePtr == invalidPage()) {
+      PagePtr = getAllocator()->template Allocate<T>(PAGE_SIZE);
+      // We need to invoke the default constructor on all the elements of the
+      // page.
+      for (std::size_t I = 0; I < PAGE_SIZE; ++I) {
+        new (PagePtr + I) T();
+      }
+    }
+    // Dereference the element in the page.
+    return *((Index % PAGE_SIZE) + PagePtr);
+  }
+
+  // Return the capacity of the vector. I.e. the maximum size it can be expanded
+  // to with the expand method without allocating more pages.
+  std::size_t capacity() const { return PageToDataIdx.size() * PAGE_SIZE; }
+
+  // Return the size of the vector. I.e. the maximum index that can be
+  // accessed, i.e. the maximum value which was used as argument of the
+  // expand method.
+  std::size_t size() const { return Size; }
+
+  // Expands the vector to the given NewSize number of elements.
+  // If the vector was smaller, allocates new pages as needed.
+  // It should be called only with NewSize >= Size.
+  void expand(std::size_t NewSize) {
```

Comment:
Should this be called `resize` (I see a bunch of places that rename `resize` calls to `expand` calls, so it sounds like it's meant to have similar/matching semantics)

Reply:
Yes, now that I use the BumpPtrAllocator it becomes somewhat easy to support shrinking. Updated with the associated code.